Allocate RuntimeType objects on Frozen Object Heap

EgorBo opened the pull request:
Follow up to https://github.com/dotnet/runtime/pull/49576
This PR allocates Type objects on the recently-added Frozen Segments so we can bake direct objects' addresses in JIT and avoid helper calls, e.g. for:
```csharp
Type Test() => typeof(int);
```
Old codegen:
```asm
; Method Program:Test():System.Type:this
G_M50870_IG01: 
       4883EC28             sub      rsp, 40
G_M50870_IG02: 
       48B968BB53F5FB7F0000 mov      rcx, 0x7FFBF553BB68      ; System.Int32
       E81D94505F           call     CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
       90                   nop      
G_M50870_IG03: 
       4883C428             add      rsp, 40
       C3                   ret      
; Total bytes of code: 25
```
New codegen:
```asm
; Method Program:Test():System.Type:this
G_M50870_IG01: 
G_M50870_IG02:          
       48B8482A63DE29020000 mov      rax, 0x229DE632A48      ; System.Int32
G_M50870_IG03: 
       C3                   ret      
; Total bytes of code: 11
```

Jit-diffs (-f --pmi):
```
PMI CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies for  default jit

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 64750613
Total bytes of diff: 64621503
Total bytes of delta: -129110 (-0.20 % of base)
Total relative delta: -715.51
    diff is an improvement.
    relative diff is an improvement.


Top file improvements (bytes):
      -14241 : System.Text.Json.dasm (-1.22% of base)
      -12215 : System.Private.Xml.dasm (-0.30% of base)
      -11900 : System.Linq.Queryable.dasm (-3.49% of base)
       -9901 : System.Private.CoreLib.dasm (-0.17% of base)
       -7878 : System.Linq.Expressions.dasm (-0.90% of base)
       -6757 : System.Data.Common.dasm (-0.39% of base)
       -6516 : System.Private.DataContractSerialization.dasm (-0.71% of base)
       -4883 : System.ComponentModel.TypeConverter.dasm (-1.59% of base)
       -4275 : Newtonsoft.Json.dasm (-0.44% of base)
       -3175 : System.ComponentModel.Composition.dasm (-0.86% of base)
       -2984 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.04% of base)
       -2625 : Microsoft.CSharp.dasm (-0.61% of base)
       -2537 : System.DirectoryServices.dasm (-0.54% of base)
       -2418 : FSharp.Core.dasm (-0.06% of base)
       -2219 : System.DirectoryServices.AccountManagement.dasm (-0.52% of base)
       -1961 : Microsoft.CodeAnalysis.CSharp.dasm (-0.04% of base)
       -1539 : System.Data.OleDb.dasm (-0.45% of base)
       -1387 : System.Drawing.Common.dasm (-0.24% of base)
       -1309 : xunit.execution.dotnet.dasm (-0.47% of base)
       -1196 : Microsoft.VisualBasic.Core.dasm (-0.22% of base)

173 total files with Code Size differences (173 improved, 0 regressed), 101 unchanged.

Top method regressions (bytes):
         122 (13.22% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[System.Numerics.Vector`1[float]](System.Object):System.Numerics.Vector`1[float]
         115 (13.92% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[int](System.Object):int
         115 (13.69% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[long](System.Object):long
         115 (13.45% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[short](System.Object):short
         115 (13.67% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[ubyte](System.Object):ubyte
         112 (11.62% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[double](System.Object):double
          99 (11.22% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[System.Nullable`1[int]](System.Object):System.Nullable`1[int]
          89 ( 5.45% of base) : System.Composition.TypedParts.dasm - System.Composition.TypedParts.ActivationFeatures.PropertyInjectionFeature:RewriteActivator(System.Reflection.TypeInfo,System.Composition.Hosting.Core.CompositeActivator,System.Collections.Generic.IDictionary`2[System.String,System.Object],System.Collections.Generic.IEnumerable`1[System.Composition.Hosting.Core.CompositionDependency]):System.Composition.Hosting.Core.CompositeActivator:this
          69 ( 8.18% of base) : System.Private.CoreLib.dasm - AdjustmentRule:.ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext):this
          68 ( 2.52% of base) : System.Linq.Expressions.dasm - System.Linq.Expressions.Interpreter.LightCompiler:CompileLiftedLogicalBinaryExpression(System.Linq.Expressions.BinaryExpression,bool):this
          62 ( 9.67% of base) : System.Data.Odbc.dasm - System.Data.Common.ADP:UnsafeCreateTimer(System.Threading.TimerCallback,System.Object,int,int):System.Threading.Timer
          59 ( 1.45% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationWriterCodeGen:WriteValue(System.Object):this
          58 ( 6.39% of base) : FSharp.Core.dasm - Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter:transBinOp(Microsoft.FSharp.Core.FSharpFunc`2[System.Type,bool],Microsoft.FSharp.Quotations.FSharpExpr,ConvEnv,bool,Microsoft.FSharp.Quotations.FSharpExpr,Microsoft.FSharp.Quotations.FSharpExpr,bool,Microsoft.FSharp.Core.FSharpFunc`2[System.Tuple`3[System.Linq.Expressions.Expression,System.Linq.Expressions.Expression,System.Reflection.MethodInfo],System.Linq.Expressions.BinaryExpression],System.RuntimeMethodHandle):System.Linq.Expressions.Expression
          56 ( 2.79% of base) : FSharp.Core.dasm - Microsoft.FSharp.Linq.RuntimeHelpers.Adapters:ConvImmutableTypeToMutableType(ConversionDescription,System.Type):System.Type
          55 ( 3.91% of base) : System.Private.CoreLib.dasm - System.Enum:TryParse[double](System.ReadOnlySpan`1[ushort],bool,bool,byref):bool
          55 ( 3.88% of base) : System.Private.CoreLib.dasm - System.Enum:TryParse[System.Numerics.Vector`1[float]](System.ReadOnlySpan`1[ushort],bool,bool,byref):bool
          54 ( 5.99% of base) : FSharp.Core.dasm - Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter:transShiftOp(Microsoft.FSharp.Quotations.FSharpExpr,ConvEnv,bool,Microsoft.FSharp.Quotations.FSharpExpr,Microsoft.FSharp.Quotations.FSharpExpr,bool,Microsoft.FSharp.Core.FSharpFunc`2[System.Tuple`3[System.Linq.Expressions.Expression,System.Linq.Expressions.Expression,System.Reflection.MethodInfo],System.Linq.Expressions.BinaryExpression],System.RuntimeMethodHandle):System.Linq.Expressions.Expression
          54 ( 4.81% of base) : System.Private.CoreLib.dasm - System.Decimal:.ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext):this
          54 ( 4.26% of base) : System.Private.CoreLib.dasm - System.Enum:TryParse[long](System.ReadOnlySpan`1[ushort],bool,bool,byref):bool
          54 ( 4.30% of base) : System.Private.CoreLib.dasm - System.Enum:TryParse[ubyte](System.ReadOnlySpan`1[ushort],bool,bool,byref):bool

Top method improvements (bytes):
       -2328 (-26.68% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.SyntaxFactory:GetNodeTypes():System.Collections.Generic.IEnumerable`1[System.Object]
       -1496 (-26.69% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxFactory:GetNodeTypes():System.Collections.Generic.IEnumerable`1[System.Type]
        -933 (-12.13% of base) : System.Data.Common.dasm - System.Data.DataTable:DeserializeTableSchema(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext,bool):this
        -714 (-12.83% of base) : System.Data.Common.dasm - System.Data.DataTable:SerializeTableSchema(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext,bool):this
        -480 (-5.29% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationReaderILGen:WriteLiteralStructMethod(System.Xml.Serialization.StructMapping):this
        -462 (-7.88% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationReaderILGen:WriteElement(System.String,System.String,System.String,System.Xml.Serialization.ElementAccessor,System.Xml.Serialization.ChoiceIdentifierAccessor,System.String,bool,bool,int,int):this
        -395 (-5.99% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationWriterILGen:WriteStructMethod(System.Xml.Serialization.StructMapping):this
        -360 (-6.65% of base) : System.ComponentModel.TypeConverter.dasm - <>c:<get_IntrinsicTypeConverters>b__24_0():System.Collections.Generic.Dictionary`2[System.Object,IntrinsicTypeConverterData]:this
        -343 (-17.03% of base) : System.Text.Json.dasm - System.Text.Json.Serialization.Converters.IEnumerableConverterFactory:CreateConverter(System.Type,System.Text.Json.JsonSerializerOptions):System.Text.Json.Serialization.JsonConverter:this
        -326 (-5.34% of base) : System.Private.DataContractSerialization.dasm - DataContractCriticalHelper:TryCreateBuiltInDataContract(System.String,System.String,byref):bool
        -305 (-6.14% of base) : System.Data.Common.dasm - System.Data.DataTableReader:GetSchemaTableFromDataTable(System.Data.DataTable):System.Data.DataTable
        -287 (-47.52% of base) : System.Linq.Expressions.dasm - System.Linq.Expressions.Compiler.DelegateHelpers:GetFuncType(System.Type[]):System.Type
        -276 (-46.78% of base) : System.Linq.Expressions.dasm - System.Linq.Expressions.Compiler.DelegateHelpers:GetActionType(System.Type[]):System.Type
        -264 (-11.64% of base) : System.Private.DataContractSerialization.dasm - DataContractCriticalHelper:GetBuiltInDataContract(System.String):System.Runtime.Serialization.DataContracts.DataContract
        -262 (-5.19% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationWriterILGen:WriteMember(System.Xml.Serialization.SourceInfo,System.Xml.Serialization.AttributeAccessor,System.Xml.Serialization.TypeDesc,System.String):this
        -240 (-6.39% of base) : System.Private.Xml.dasm - System.Xml.Serialization.SchemaGraph:Depends(System.Xml.Schema.XmlSchemaObject,System.Collections.ArrayList):this
        -233 (-6.52% of base) : Microsoft.CSharp.dasm - Microsoft.CSharp.RuntimeBinder.ComInterop.ComInvokeBinder:GenerateTryBlock():System.Linq.Expressions.Expression:this
        -232 (-6.14% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationWriterILGen:WriteEnumMethod(System.Xml.Serialization.EnumMapping):this
        -228 (-5.30% of base) : System.Private.Xml.dasm - System.Xml.Serialization.XmlSerializationWriterILGen:WriteEnumAndArrayTypes():this
        -227 (-5.05% of base) : System.Private.Xml.dasm - System.Xml.Serialization.ReflectionAwareCodeGen:WriteTypeInfo(System.Xml.Serialization.TypeScope,System.Xml.Serialization.TypeDesc,System.Type):System.String:this

Top method regressions (percentages):
          35 (14.11% of base) : FSharp.Core.dasm - Microsoft.FSharp.Text.StructuredPrintfImpl.Display:isSetOrMapType(System.Type):bool
         115 (13.92% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[int](System.Object):int
         115 (13.69% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[long](System.Object):long
         115 (13.67% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[ubyte](System.Object):ubyte
         115 (13.45% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[short](System.Object):short
         122 (13.22% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[System.Numerics.Vector`1[float]](System.Object):System.Numerics.Vector`1[float]
          11 (11.83% of base) : System.Composition.Hosting.dasm - System.Composition.Hosting.Util.MethodInfoExtensions:CreateStaticDelegate[System.Numerics.Vector`1[float]](System.Reflection.MethodInfo):System.Numerics.Vector`1[float]
         112 (11.62% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[double](System.Object):double
          99 (11.22% of base) : Microsoft.VisualBasic.Core.dasm - Microsoft.VisualBasic.CompilerServices.Conversions:ToGenericParameter[System.Nullable`1[int]](System.Object):System.Nullable`1[int]
          62 ( 9.67% of base) : System.Data.Odbc.dasm - System.Data.Common.ADP:UnsafeCreateTimer(System.Threading.TimerCallback,System.Object,int,int):System.Threading.Timer
          44 ( 9.17% of base) : Microsoft.Extensions.DependencyInjection.Abstractions.dasm - Microsoft.Extensions.DependencyInjection.ActivatorUtilities:CreateFactory(System.Type,System.Type[]):Microsoft.Extensions.DependencyInjection.ObjectFactory
          69 ( 8.18% of base) : System.Private.CoreLib.dasm - AdjustmentRule:.ctor(System.Runtime.Serialization.SerializationInfo,System.Runtime.Serialization.StreamingContext):this
          51 ( 7.92% of base) : FSharp.Core.dasm - Microsoft.FSharp.Linq.RuntimeHelpers.LeafExpressionConverter:transBoolOpNoWitness(Microsoft.FSharp.Core.FSharpFunc`2[System.Type,bool],ConvEnv,bool,Microsoft.FSharp.Quotations.FSharpExpr,Microsoft.FSharp.Quotations.FSharpExpr,bool,Microsoft.FSharp.Core.FSharpFunc`2[System.Tuple`4[System.Linq.Expressions.Expression,System.Linq.Expressions.Expression,bool,System.Reflection.MethodInfo],System.Linq.Expressions.BinaryExpression],System.Reflection.MethodInfo):System.Linq.Expressions.Expression
          40 ( 7.60% of base) : System.Composition.TypedParts.dasm - System.Composition.TypedParts.Discovery.DiscoveredPropertyExport:GetExportDescriptor(System.Composition.Hosting.Core.CompositeActivator):System.Composition.Hosting.Core.ExportDescriptor:this
          42 ( 7.24% of base) : Newtonsoft.Json.dasm - Newtonsoft.Json.Utilities.ExpressionReflectionDelegateFactory:CreateMethodCall[System.__Canon](System.Reflection.MethodBase):Newtonsoft.Json.Utilities.MethodCall`2[System.__Canon,System.__Canon]:this
          48 ( 6.69% of base) : Newtonsoft.Json.dasm - Newtonsoft.Json.Utilities.ExpressionReflectionDelegateFactory:CreateSet[double](System.Reflection.FieldInfo):System.Action`2[double,System.Object]:this
          48 ( 6.69% of base) : Newtonsoft.Json.dasm - Newtonsoft.Json.Utilities.ExpressionReflectionDelegateFactory:CreateSet[int](System.Reflection.FieldInfo):System.Action`2[int,System.Object]:this
          48 ( 6.69% of base) : Newtonsoft.Json.dasm - Newtonsoft.Json.Utilities.ExpressionReflectionDelegateFactory:CreateSet[long](System.Reflection.FieldInfo):System.Action`2[long,System.Object]:this
          48 ( 6.69% of base) : Newtonsoft.Json.dasm - Newtonsoft.Json.Utilities.ExpressionReflectionDelegateFactory:CreateSet[short](System.Reflection.FieldInfo):System.Action`2[short,System.Object]:this
          48 ( 6.69% of base) : Newtonsoft.Json.dasm - Newtonsoft.Json.Utilities.ExpressionReflectionDelegateFactory:CreateSet[System.Nullable`1[int]](System.Reflection.FieldInfo):System.Action`2[System.Nullable`1[int],System.Object]:this

Top method improvements (percentages):
          -8 (-88.89% of base) : System.Diagnostics.DiagnosticSource.dasm - System.Diagnostics.Metrics.Instrument:ValidateTypeParameter[double]()
          -8 (-88.89% of base) : System.Diagnostics.DiagnosticSource.dasm - System.Diagnostics.Metrics.Instrument:ValidateTypeParameter[int]()
          -8 (-88.89% of base) : System.Diagnostics.DiagnosticSource.dasm - System.Diagnostics.Metrics.Instrument:ValidateTypeParameter[long]()
          -8 (-88.89% of base) : System.Diagnostics.DiagnosticSource.dasm - System.Diagnostics.Metrics.Instrument:ValidateTypeParameter[short]()
          -8 (-88.89% of base) : System.Diagnostics.DiagnosticSource.dasm - System.Diagnostics.Metrics.Instrument:ValidateTypeParameter[ubyte]()
         -40 (-71.43% of base) : System.Linq.Expressions.dasm - System.Linq.Expressions.Expression:IfThenElse(System.Linq.Expressions.Expression,System.Linq.Expressions.Expression,System.Linq.Expressions.Expression):System.Linq.Expressions.ConditionalExpression
         -23 (-62.16% of base) : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.TypeHandle:TypeHandleOf[double]():System.Runtime.CompilerServices.TypeHandle
         -23 (-62.16% of base) : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.TypeHandle:TypeHandleOf[int]():System.Runtime.CompilerServices.TypeHandle
         -23 (-62.16% of base) : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.TypeHandle:TypeHandleOf[long]():System.Runtime.CompilerServices.TypeHandle
         -23 (-62.16% of base) : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.TypeHandle:TypeHandleOf[short]():System.Runtime.CompilerServices.TypeHandle
         -23 (-62.16% of base) : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.TypeHandle:TypeHandleOf[System.Nullable`1[int]]():System.Runtime.CompilerServices.TypeHandle
         -23 (-62.16% of base) : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.TypeHandle:TypeHandleOf[System.Numerics.Vector`1[float]]():System.Runtime.CompilerServices.TypeHandle
         -23 (-62.16% of base) : System.Private.CoreLib.dasm - System.Runtime.CompilerServices.TypeHandle:TypeHandleOf[ubyte]():System.Runtime.CompilerServices.TypeHandle
         -24 (-61.54% of base) : System.Private.CoreLib.dasm - System.Runtime.InteropServices.Marshal:DestroyStructure[double](long)
         -24 (-61.54% of base) : System.Private.CoreLib.dasm - System.Runtime.InteropServices.Marshal:DestroyStructure[int](long)
         -24 (-61.54% of base) : System.Private.CoreLib.dasm - System.Runtime.InteropServices.Marshal:DestroyStructure[long](long)
         -24 (-61.54% of base) : System.Private.CoreLib.dasm - System.Runtime.InteropServices.Marshal:DestroyStructure[short](long)
         -24 (-61.54% of base) : System.Private.CoreLib.dasm - System.Runtime.InteropServices.Marshal:DestroyStructure[System.Nullable`1[int]](long)
         -24 (-61.54% of base) : System.Private.CoreLib.dasm - System.Runtime.InteropServices.Marshal:DestroyStructure[System.Numerics.Vector`1[float]](long)
         -24 (-61.54% of base) : System.Private.CoreLib.dasm - System.Runtime.InteropServices.Marshal:DestroyStructure[ubyte](long)

7846 total methods with Code Size differences (7511 improved, 335 regressed), 383629 unchanged.
```
Size regressions look to be CSE-related, e.g. https://www.diffchecker.com/eOr1LIyf
E.g. previously `call CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE` could share its const argument with other places where we only need type handle (like in this diff)

Notes:
In JIT I do the optimization in morph because a lot of other optimizations here and there (mostly in importer) rely on `typeof()` being a helper call so it's simpler for now, will try to perform a sort of clean up later - current change is minimalistic and doesn't produce regressions.

Closes https://github.com/dotnet/runtime/issues/49429

EgorBo committed:
Allocate Type objects on Frozen Heap

EgorBo committed:
Mitigate perf regressions

EgorBo committed:
clean up

EgorBo committed:
fix build issue?

EgorBo committed:
Fix eeGetCPString for string literals in JitDisasm

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into foh-type-objects

EgorBo committed:
clean up

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into foh-type-objects

EgorBo commented on src/coreclr/jit/emit.cpp:
```diff
@@ -4129,42 +4129,15 @@ void emitter::emitDispCommentForHandle(size_t handle, size_t cookie, GenTreeFlag
     const char* str = nullptr;
     if (flag == GTF_ICON_STR_HDL)
+    {
+        str = "string handle";
```
I don't print comment for `GTF_ICON_STR_HDL` because it's expected to be rare now (string literals are mostly frozen) and didn't want to keep `eeGetCPString` for it because that is a bit dangerous and contained a gc hole

jakobbotsch replied:
I don't think there was a GC hole since the users of `eeGetCPString` do not call any JIT-EE functions between reading the handle and reading the characters, so no GC could occur.

jkotas replied:
I do not think that this runs in coop mode. JITing is a long running piece of unmanaged code. It always runs in preemptive mode to avoid blocking GC.

jakobbotsch replied:
Ah okay, so there is a GC hole. Sorry for the misinformation.

EgorBo commented:
Seems to be passing CI tests, @jkotas when you have time could you review the vm side?

EgorBo commented on src/coreclr/vm/frozenobjectheap.cpp:
```diff
@@ -85,6 +85,15 @@ Object* FrozenObjectHeapManager::TryAllocateObject(PTR_MethodTable type, size_t
+bool FrozenObjectHeapManager::IsFromFrozenSegment(Object* object)
+{
+#ifdef FEATURE_BASICFREEZE
+    return GCHeapUtilities::GetGCHeap()->IsInFrozenSegment(object);
```
Alternatively I can enumerate segments in FrozenObjectHeapManager and check if the pointer belongs to any

jkotas replied:
This method won't be necessary if you incorporate my other feedback.

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4222,9 +4223,20 @@ OBJECTREF MethodTable::GetManagedClassObject()
+        {
+            refClass = (REFLECTCLASSBASEREF)AllocateObject(g_pRuntimeTypeClass);
+        }
 
```
We should do the same thing as what we have done for string literals: Skip the handle allocation for frozen objects and use a bit on writeable data to tell whether it is frozen object or not.

EgorBo replied:
@jkotas I am not sure I fully understand - writeable data seems to be 8 bytes (on x64) just for the pointer to type object, how can I extend it to include extra bits? Did you mean to use some bit in dwFlags?

EgorBo replied:
So is `gc` struct

```
    struct
    {
        OBJECTREF value;
        LOADERALLOCATORREF loaderAllocator;
        PTRARRAYREF handleTable;
        PTRARRAYREF handleTableOld;
    } gc;
```
seems to be packed?

EgorBo replied:
Ah, `loaderAllocator` is unused most of the time like in string handles if I read it correctly

EgorBo replied:
@jkotas Sorry for pinging again, but I've tried again and it feels like I am a bit lost

so what we save into `MethodTable::m_hExposedClassObject` is a pointer-sized slot in a pinned table that is pointing to `RuntimeType` so I am bit lost on where exactly I can save that bit. What's even more difficult to understand for me is why we [add](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/loaderallocator.cpp#L844) `+ 1` here and then [subtract `1`](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/loaderallocator.inl#L98) when we obtain `RuntimeType` from that slot - so basically why it's pointing to the next slot?

jkotas replied:
I would help to make the frozen objects mandatory to reduce the complexity.

If we make the frozen objects mandatory, the logic will become pretty simple: if the low bit is not set, `m_hExposedClassObject` would point directly at the frozen RuntimeTypeHandle (ie we would save the indirection compare to what we have today). If the low bit is set, `m_hExposedClassObject` would point at the slot in collectible handle table like it does today.

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4222,9 +4223,20 @@ OBJECTREF MethodTable::GetManagedClassObject()
+        }
 
         ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetType(TypeHandle(this));
         ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetKeepAlive(pLoaderAllocator->GetExposedObject());
```
Setting the KeepAlive object should be only done for the non-frozen case, and we may want to assert that `pLoaderAllocator->GetExposedObject()` is null for the frozen case.

MichalStrehovsky commented on src/coreclr/jit/compiler.h:
```diff
@@ -6808,7 +6809,7 @@ class Compiler
 #define OMF_HAS_PATCHPOINT                     0x00000100 // Method contains patchpoints
 #define OMF_NEEDS_GCPOLLS                      0x00000200 // Method needs GC polls
-#define OMF_HAS_FROZEN_STRING                  0x00000400 // Method has a frozen string (REF constant int), currently only on NativeAOT.
+#define OMF_HAS_FROZEN_OBJECTS                 0x00000400 // Method has a frozen string (REF constant int), currently only on NativeAOT.
```
```suggestion
#define OMF_HAS_FROZEN_OBJECTS                 0x00000400 // Method has a frozen string (REF constant int)
```

MichalStrehovsky commented on src/coreclr/jit/morph.cpp:
```diff
@@ -8312,6 +8312,26 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
+    {
+        GenTree*             argNode = call->AsCall()->gtArgs.GetArgByIndex(0)->GetNode();
+        CORINFO_CLASS_HANDLE hClass  = gtGetHelperArgClassHandle(argNode);
+        if (!opts.IsReadyToRun() && (hClass != NO_CLASS_HANDLE) && !gtIsActiveCSE_Candidate(argNode))
```
What are all the places that make this incompatible with ReadyToRun compilation (NativeAOT also returns true for this).

I see the `CORINFO_RuntimeType` where JIT is now making assumptions about how RuntimeType looks like. But that's just to get a name. We could have a dedicated JitInterface API for that and keep it a handle.

Frozen RuntimeType instances can be supportable on NativeAOT side as well if RyuJIT treats RuntimeTypes like any other handles and doesn't do the CORINFO_RuntimeType with actual fields.

EgorBo replied:
`CORINFO_RuntimeType` is only used to get a string name for a type in JitDisasm comments - it's probably better to give up on it than introducing a new JIT-EE interface 😄

jkotas replied:
Can you use `GenTreeIntCon::gtCompileTimeHandle` for this purpose? Initialize it to point to CORINFO_CLASS_HANDLE, etc. and then get the name from the CORINFO_CLASS_HANDLE using existing JIT/EE interface method?

jkotas commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1615,45 +1615,45 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
-
-    // make this bulletproof, so it works even if we are wrong.
-    if (ReadProcessMemory(GetCurrentProcess(), (void*)strHandle, buff, 4, nullptr) == 0)
+    auto typeOfHandle = *((CORINFO_CLASS_HANDLE*)handle);
```
This works for the in-process JIT, but it is broken and can crash for everything else (super PMI, crossgen, nativeaot).

All this logic should be behind JIT/EE interface API as Michal said.

MichalStrehovsky commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4222,9 +4223,20 @@ OBJECTREF MethodTable::GetManagedClassObject()
+            // Allocate RuntimeType on a frozen segment
+            FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
+            size_t objSize = g_pRuntimeTypeClass->GetBaseSize();
+            refClass = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(foh->TryAllocateObject(g_pRuntimeTypeClass, objSize));
```
I think this memory is leaked if the InterlockedCompareExchangeT below fails. Might be okay.

jkotas replied:
It may be a good idea to just take a lock around this so that there is no leak.

TryAllocateObject may take a lock, and so you can end up in a pathological situations where a bunch of threads get stacked up behind this lock and all of them leak.

This leak is not a problem for non-frozen instances. The GC will clean them up.

EgorBo committed:
Update src/coreclr/jit/compiler.h

Co-authored-by: Michal Strehovský <MichalStrehovsky@users.noreply.github.com>

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into foh-type-objects

# Conflicts:
#	src/coreclr/inc/jiteeversionguid.h
#	src/coreclr/jit/valuenum.cpp

EgorBo committed:
Address some of the feedback [WIP]

EgorBo committed:
Address feedback

EgorBo committed:
Fix compilation

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4220,19 +4221,34 @@ OBJECTREF MethodTable::GetManagedClassObject()
+            // Allocate RuntimeType on a frozen segment
+            FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
+            size_t objSize = g_pRuntimeTypeClass->GetBaseSize();
+            Object* obj = foh->TryAllocateObject(g_pRuntimeTypeClass, objSize);
```
```suggestion
            Object* obj = foh->TryAllocateObject(g_pRuntimeTypeClass, objSize);
            _ASSERTE(obj != NULL);
```

EgorBo committed:
Introduce `objectToString` JIT-EE API

EgorBo committed:
Update src/coreclr/vm/methodtable.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

EgorBo committed:
check for unloadability in GetPinnedManagedClassObjectIfExists

EgorBo committed:
Merge branch 'foh-type-objects' of github.com:EgorBo/runtime-1 into foh-type-objects

EgorBo committed:
Since I still have to check CanUnload I guess I don't need the bit test

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -5945,6 +5987,40 @@ CorInfoHelpFunc CEEInfo::getUnBoxHelper(CORINFO_CLASS_HANDLE clsHnd)
+    GCX_COOP();
+
+    TypeHandle typeHnd(clsHnd);
+    if (!typeHnd.IsNull() && !typeHnd.IsTypeDesc() && typeHnd.IsFullyLoaded())
```
You should be able to do this optimization for TypeDescs too.

Also, IsFullyLoaded check should not be needed. Types handed back to the JIT are always fully loaded.

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4220,19 +4221,31 @@ OBJECTREF MethodTable::GetManagedClassObject()
-        LOADERHANDLE exposedClassObjectHandle = pLoaderAllocator->AllocateHandle(refClass);
-
         if (InterlockedCompareExchangeT(&GetWriteableDataForWrite()->m_hExposedClassObject, exposedClassObjectHandle, static_cast<LOADERHANDLE>(NULL)))
         {
```
This backout if not correct.

Do you plan to implement the lock mentioned in the other comment?

EgorBo replied:
Ah, yes, it was supposed to be lock, I was just not sure where to define Crst object

jkotas commented on src/coreclr/vm/methodtable.inl:
```diff
@@ -1342,12 +1359,14 @@ FORCEINLINE OBJECTREF MethodTable::GetManagedClassObjectIfExists()
 
-    OBJECTREF retVal;
+    // First, check if have a cached reference to an effectively pinned (allocated on FOH) object
+    OBJECTREF retVal = GetPinnedManagedClassObjectIfExists();
```
`GetManagedClassObjectIfExists` if perf critical method. It is used by `JIT_GetRuntimeType` that is still going to be used by unloadable and shared generic code. 

You want to make sure that it has same or better codegen compared to what it is today. I think it is significantly worse in the current state (`CanUnload` called by `GetPinnedManagedClassObjectIfExists` is virtual method call).

EgorBo replied:
Ah, good point, turns out the idea with the lowest bit is doable - switched to that

jkotas commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -1815,6 +1815,14 @@ private int getStringLiteral(CORINFO_MODULE_STRUCT_* module, uint metaTOK, char*
+        private int objectToString(void* handle, char* buffer, int size)
+#pragma warning restore CA1822 // Mark members as static
+        {
+            // TODO: implement
```
`var obj = HandleToObject((IntPtr)handle);` will give you the compiler object, and the rest should be simple.

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4220,19 +4221,31 @@ OBJECTREF MethodTable::GetManagedClassObject()
-
-        ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetType(TypeHandle(this));
-        ((ReflectClassBaseObject*)OBJECTREFToObject(refClass))->SetKeepAlive(pLoaderAllocator->GetExposedObject());
+        LOADERHANDLE exposedClassObjectHandle = NULL;
```
This is not a technically `LOADERHANDLE` since it does not follow the `LOADERHANDLE` contract anymore.

Use a different name (e.g. like `RUNTIMETYPEHANDLE`) to avoid confusion?

EgorBo committed:
Address feedback

EgorBo committed:
Add more temp debug checks for ci

jkotas commented on src/coreclr/vm/vars.hpp:
```diff
@@ -99,6 +99,8 @@ struct LOADERHANDLE__
 };
```
Nit: You can delete `struct LOADERHANDLE__`. Looks like left-over from earlier definition of `LOADERHANDLE`.

EgorBo commented:
CI failures are weird, e.g. https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-75573-merge-3616944558404cd193/System.Linq.Queryable.Tests/1/console.13d74af2.log?helixlogtype=result had no luck trying to reproduce any, going to kick off a single gc stress job - maybe it will help to find a repro.

EgorBo commented:
/azp run Fuzzlyn

EgorBo commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

EgorBo committed:
Check if it reproduces because of jit

EgorBo committed:
Check if it reproduces because of jit

EgorBo commented:
Ok, gcstress=0xc didn't find anything, and it doesn't look like it's a codegen issue (I disabled the jit optimization) 
Reproduces for me locally (not every run) for `src\libraries\System.Formats.Asn1\tests` built with Debug.

Judging by
```
System.ArrayTypeMismatchException: Attempted to access an element as a type incompatible with the array
```
it feels like it's related to type co(ntra)variance

EgorBo committed:
Ah, facepalm

EgorBo committed:
Clean up

EgorBo commented:
@jkotas Anything else needed from VM side? It seems to be passing all tests (spmi jobs are failing due to JIT-EE guid change), I'll kick off (_gently this time_) a few outerloop pipelines to make sure. There are a few things I'd love to address separately later:
1) Do the same for TypeDesc (diffs are pretty small from it though)
2) Wire it up in NativeAOT (or can leave it for anyone interested)
3) `PublishObjectAndNotify` problem from https://github.com/dotnet/runtime/pull/49576#issuecomment-1250248951

EgorBo commented:
https://github.com/icsharpcode/AvaloniaILSpy - when I run this average-size client app it allocates 12153 objects on the frozen segments (total size is 604kb). Going to check other apps/web services

jkotas commented:
> Do the same for TypeDesc (diffs are pretty small from it though)

I think it would be a good idea to address this as part of this PR. (I expect that you will want to move some the code that you are adding here around.)

The rest can wait for separate PR.

EgorBo committed:
Address Jan's feedback

EgorBo commented:
> > Do the same for TypeDesc (diffs are pretty small from it though)
> 
> I think it would be a good idea to address this as part of this PR. (I expect that you will want to move some the code that you are adding here around.)
> 
> The rest can wait for separate PR.

Ok, pushed the change for `ParamTypeDesc/TypeVarTypeDesc`. The only thing I am not sure is the fact they all share the same lock. From the other hand, FrozenObjectHeapManager has its own lock they all will hit anyway if I create 3 different lock objects

EgorBo committed:
Oops, a typo

EgorBo commented:
Ah, oops forgot to update the logic to extract OBJECTREF from the field for both

EgorBo committed:
Share logic to extract OBJECTREF from m_hExposedClassObject

EgorBo commented:
/azp run runtime-coreclr outerloop

jkotas commented on src/coreclr/vm/methodtable.inl:
```diff
@@ -1335,25 +1335,17 @@ inline OBJECTHANDLE MethodTable::GetLoaderAllocatorObjectHandle()
-    COMPILER_ASSUME(retVal != NULL);
-    return retVal;
+//==========================================================================================
+FORCEINLINE OBJECTREF MethodTable::GetManagedClassObjectIfExists()
```
This still needs some fine-tunning. 

Run this:
```
using System.Diagnostics;
using System.Runtime.CompilerServices;

Stopwatch sw = new();
for (;;)
{
   sw.Restart();
   for (int i = 0; i < 1_000_000_000; i++) MyG<string>.M();
   Console.WriteLine(sw.ElapsedMilliseconds);
}

class MyG<T>
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static Type M() => typeof(T);
}
```
It has more than 2x regression compared to main with this change.

If you do this right, it should be a bit faster compared to main due to one less indirection.

EgorBo replied:
@jkotas Thanks, will take a look but I was hoping on Native PGO to kick in and reshape it

jkotas replied:
I not sure whether native PGO would be able to clean this up completely. In any case, it would be unreliable at best.

EgorBo replied:
Dissassembly for `JIT_GetRuntimeType`
### Main (compiled with `/p:NoPgoOptimize=true`):
![image](https://user-images.githubusercontent.com/523221/191016935-dacbed23-48f4-4aa0-bda9-5dd543868024.png)

### This PR (also with `/p:NoPgoOptimize=true`):
![image](https://user-images.githubusercontent.com/523221/191017034-84059cbc-5557-4626-8a22-e75a3daa92b0.png)

EgorBo replied:
Experimenting with it to produce better asm now

EgorBo replied:
Pushed [a small change](https://github.com/dotnet/runtime/pull/75573/commits/e359e66fe0713eac59a6fb469247f3b1a3f0939c) that drastically improve asm for `JIT_GetRuntimeType` - I checked few times. Now idea why MSVC emitted such a bad code for the previous version.

![image](https://user-images.githubusercontent.com/523221/191021602-62b76bf4-3ffb-47f9-8151-faaad3337508.png)

EgorBo replied:
Ok, did some massage in C++ to produce better code here is what I ended up with: 
![image](https://user-images.githubusercontent.com/523221/191035497-a58228cd-2e55-40e2-a04c-ccf8a497e0a3.png)

Let me benchmark it on your benchmark

EgorBo replied:
![image](https://user-images.githubusercontent.com/523221/191036183-0406e40a-e2be-492b-bd3c-db35ffb1200a.png)

runtime-main - this PR, runtime7 - Main (the same commit PR is based on)
both were compiled with Release + `/p:NoPgoOptimize=true`

Disabled ReadyToRun to turn off managed static PGO just in case. Main is on noise-level faster but presumably Native PGO will make PR better as was highlighted on the screen ^

jkotas replied:
> we have to do "test rdx, rdx"

We can keep the polarity of the low-bit to avoid this conditional jump.

> redundant "test rdx,rdx"

It is not redundant on some paths given the way the code is written. This should be fixable by structuring the code differently, similar to how it was structured before.

EgorBo replied:
> It is not redundant on some

Right, that's what I meant by "Doesn't MSVC support jump-threading, a compiler optimization that is supposed to handle such cases"

> We can keep the polarity of the low-bit to avoid this conditional jump.

@jkotas do you mean to set the lowest bit for gc ref instead? - is it safe to modify a bit in a gc reference like that?

EgorBo replied:
oh I see what you mean but then I would need to clear it before returning to JIT so we'll spend cycles there instead while currently I don't even need to clear it

jkotas replied:
It is always better to spend a few extra instructions during JITing than to spend a few extra instruction when executing the JITed method.

It should work nicely together if you take https://github.com/dotnet/runtime/pull/75573#discussion_r974841009.

EgorBo commented:
/azp runtime-coreclr jitstressregs

MichalStrehovsky commented:
> Wire it up in NativeAOT (or can leave it for anyone interested)

We'll need to do some work on the reflection stack first - there's no unique `RuntimeType` type we could preallocate (NativeAOT has like half a dozen RuntimeType descendants), and it has GC-pointer fields (no-no in frozen segments). Having a single `RuntimeType` is our plan. It's just not there yet.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into foh-type-objects

EgorBo committed:
Produce better asm code for JIT_GetRuntimeType

EgorBo commented:
@jakobbotsch @AndyAyersMS PTAL JIT side, like @jakobbotsch suggested I introduced a separate flag `GTF_ICON_OBJ_HDL` for both frozen strings and frozen RuntimeType objects. `GTF_ICON_STR_HDL` is kept for non-frozen string literals (e.g. too large or in unloadable context). Also, removed dangerous `eeGetCPString` and replaced it with JIT-EE calls (tested with SPMI)

Also note that I do the optimization in Morph (fgMorphCall) and not earlier because a lot of code in JIT expects `call helper(handle)` for optimizations during import/inlining - that's how I avoided regressions and minimized changes in JIT, will do a clean up at some point.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -5945,6 +5987,40 @@ CorInfoHelpFunc CEEInfo::getUnBoxHelper(CORINFO_CLASS_HANDLE clsHnd)
+    GCX_COOP();
+
+    TypeHandle typeHnd(clsHnd);
+    if (!typeHnd.IsNull() && !typeHnd.IsTypeDesc())
```
You have fixed the typedescs. Why not take advantage of it here?

Also, I would omit the null check. The JIT should not be passing null type handles to JIT/EE interface.

EgorBo replied:
Makes sense, although my app I tested this PR on (~100k methods to jit) hit this path with TypeDesc only once

EgorBo replied:
I'll have to apply this patch https://github.com/EgorBo/runtime-1/commit/908dc940df6201658d124624dd7373df2cd82fd5
but considering minimal impact on diff perhaps it's better to keep as is?

jkotas replied:
You should not need to duplicate the logic for getting the managed class object. Can it look like this?

```
void* CEEInfo::getRuntimeTypePointer(CORINFO_CLASS_HANDLE clsHnd)
{
    CONTRACTL{
        THROWS;
        GC_TRIGGERS;
        MODE_PREEMPTIVE;
    } CONTRACTL_END;

    void* pointer = nullptr;

    JIT_TO_EE_TRANSITION();

    TypeHandle typeHnd(clsHnd);
    if (!typeHnd.IsCanonicalSubtype() && typeHnd.IsManagedClassObjectPinned())
    {
        GCX_COOP();
        pointer = OBJECTREFToObject(typeHnd.GetManagedClassObject());
    }
    
    EE_TO_JIT_TRANSITION();

    return pointer;
}

bool TypeHandle::IsManagedClassObjectPinned()
{
    LIMITED_METHOD_CONTRACT;

    // Function pointers are always mapped to typeof(IntPtr)
    return !GetLoaderAllocator()->CanUnload() || IsFnPtrType();
}

// The AllocateRuntimeTypeObject helper method can live here as well. MethodTable is probably not the best place for it.

void TypeHandle::AllocateManagedClassObject(RUNTIMETYPEHANDLE* pDest)
{
    CONTRACTL {
        THROWS;
        GC_TRIGGERS;
        MODE_COOPERATIVE;
    } CONTRACTL_END;

    LoaderAllocator* pLoaderAllocator = GetLoaderAllocator();
    if (!pLoaderAllocator->CanUnload())
    {
...
    }
    else
    {
...
    }
}
```

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4219,28 +4220,92 @@ OBJECTREF MethodTable::GetManagedClassObject()
+        refClass->SetKeepAlive(allocator->GetExposedObject());
+        LOADERHANDLE exposedClassObjectHandle = allocator->AllocateHandle(refClass);
+        // Set the lowest bit, we use it GetPinnedManagedClassObjectIfExists as a fast detection of the unloadable context
+        exposedClassObjectHandle |= 1;
```
If you set the bit here, you would need to clear the bit on the backout path. This problem will go away if you incorporate my other feedback.

EgorBo replied:
@jkotas could you please point me to that feedback, I am a bit lost.
the bit here is set for objects in unloadable contexts

jkotas replied:
"We can keep the polarity of the low-bit to avoid this conditional jump."  from https://github.com/dotnet/runtime/pull/75573#discussion_r974757277

EgorBo replied:
From my understanding, I don't need to clear it - it's ignored here https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/loaderallocator.inl#L127 when it gets index out of it

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into foh-type-objects

jakobbotsch commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1615,45 +1615,19 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
-    if (ReadProcessMemory(GetCurrentProcess(), (void*)strHandle, buff, 4, nullptr) == 0)
+    const int maxStrSize      = 128;
+    char16_t  str[maxStrSize] = {0};
+    int       realLength      = this->info.compCompHnd->objectToString((void*)handle, str, maxStrSize);
```
I think this should have an SPMI error trap around it.

jakobbotsch replied:
Actually I can see that `repObjectToString` does not fail on missing information so seems fine.

jakobbotsch commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -715,6 +715,48 @@ int CEEInfo::getStringLiteral (
+    if (stackStr.GetCount() > 0)
+    {
+        result = min(bufferSize, (int)stackStr.GetCount());
+        memcpy((BYTE*)buffer, stackStr.GetUnicode(), result * 2);
```
This does not null terminate the string, does it? JIT is assuming the string is null terminated.

Same problem in crossgen2.

EgorBo replied:
Thanks! fixed

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into foh-type-objects

EgorBo committed:
Address feedback

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -5945,6 +5991,27 @@ CorInfoHelpFunc CEEInfo::getUnBoxHelper(CORINFO_CLASS_HANDLE clsHnd)
+/***********************************************************************/
+void* CEEInfo::getRuntimeTypePointer(CORINFO_CLASS_HANDLE clsHnd)
+{
+    LIMITED_METHOD_CONTRACT;
```
```suggestion
    CONTRACTL{
        THROWS;
        GC_TRIGGERS;
        MODE_PREEMPTIVE;
    } CONTRACTL_END;
```
`LIMITED_METHOD_CONTRACT;` means that the method does not throw exceptions, allocates or switches GC mode. This does all of it.

jkotas commented on src/coreclr/vm/typehandle.cpp:
```diff
@@ -343,6 +344,93 @@ BOOL TypeHandle::IsCanonicalSubtype() const
+    return !GetLoaderAllocator()->CanUnload() || IsFnPtrType();
+}
+
+void TypeHandle::AllocateManagedClassObject(LoaderAllocator* allocator, RUNTIMETYPEHANDLE* pDest, TypeHandle type)
```
You do not need to pass loader allocator as separate argument. It is easy to get it from TypeHandle inside the implementation.

EgorBo replied:
Made it an instance method of TypeHandle (or it's too much?) 🙂

EgorBo committed:
Update src/coreclr/vm/jitinterface.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

EgorBo committed:
Address feedback

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4219,26 +4219,7 @@ OBJECTREF MethodTable::GetManagedClassObject()
-        }
-
-        GCPROTECT_END();
+        TypeHandle(this).AllocateManagedClassObject((RUNTIMETYPEHANDLE*)&GetWriteableData()->m_hExposedClassObject);
```
```suggestion
        TypeHandle(this).AllocateManagedClassObject(&GetWriteableData()->m_hExposedClassObject);
```
This cast looks unnecessary

EgorBo replied:
Ah, it turns out it's needed to drop `const RUNTIMETYPEHANDLE*`

EgorBo committed:
Update src/coreclr/vm/methodtable.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

EgorBo commented:
Current codegen for `JIT_GetRuntimeType`:

![image](https://user-images.githubusercontent.com/523221/191385711-04554837-a263-4f10-ac9f-09da1929be53.png)

one less branch but + `and     rdx, 0FFFFFFFFFFFFFFFEh` so presumably it's faster. The fast-return block is still too far but looks like MSVC just wants to have a single epilogue, I do hope it will change with native PGO 🙄

jkotas commented on src/coreclr/vm/typehandle.cpp:
```diff
@@ -343,6 +344,99 @@ BOOL TypeHandle::IsCanonicalSubtype() const
+    if (handle & 1)
+    {
+        // Clear the "is pinned object" bit from the managed reference
+        return (OBJECTREF)((handle >> 1) << 1); // C++ compiler is expected to emit "and reg, mask"
```
Why is `(x >> 1) << 1` and hoping for an optimization to kick in better, than doing ` - 1` like what the existing code does?

EgorBo replied:
Ah, indeed! with `-1` it's now `lea     rax, [rdx-1]` instead of `and rdx, 0FFFFFFFFFFFFFFFEh`

EgorBo commented:
Btw, for generics we can probably inline a fast path in jit, e.g. for:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static Type Foo<T>() => typeof(T);
```
for `__Canon` version we currently emit:
```asm
       488B4938             mov      rcx, qword ptr [rcx+38H]
       488B09               mov      rcx, qword ptr [rcx]
       E84BD8505F           call     CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
```
while we can inline the `m_Exposed & 1` check if it's worth the effort

EgorBo committed:
Address feedback: the bit is always there so we can do ` - 1`

jkotas commented:
> The fast-return block is still too far but looks like MSVC just wants to have a single epilogue, I do hope it will change with native PGO

It won't. It is impossible for the C++ compiler to prove that `handle - 1` is non-zero so it has to produce the redundant check.

I think there is a reason why the signature of `LoaderAllocator::GetHandleValueFast` is `bool (handle, objectref*)` and not `objectref (handle)`. If you change the signature of `TypeHandle::GetManagedClassObjectFromHandleFast` to be `bool (handle, objectref*)`, you should be able to get rid of the manual inlining and the redundant check.

EgorBo committed:
Address feedback

EgorBo commented:
@jkotas thanks!! now it's clear why the initial shape was like that. I had to also play with `GetManagedClassObjectFromHandleFast` because it didn't want to inline.

The new codegen is perfect:

![image](https://user-images.githubusercontent.com/523221/191391117-5e82dc89-0d45-4828-b916-279a6e672d26.png)

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -4219,26 +4219,7 @@ OBJECTREF MethodTable::GetManagedClassObject()
-        }
-
-        GCPROTECT_END();
+        TypeHandle(this).AllocateManagedClassObject((RUNTIMETYPEHANDLE*)&GetWriteableData()->m_hExposedClassObject);
```
```suggestion
        TypeHandle(this).AllocateManagedClassObject(&GetWriteableDataForWrite()->m_hExposedClassObject);
```
What about this?

jkotas reviewed (approved):
VM changes LGTM. Thanks!

jkotas committed:
Update src/coreclr/vm/methodtable.cpp

EgorBo commented:
> VM changes LGTM. Thanks!

Thanks for patience and great help! Learned so many cool tricks and VM internals! 🙂 
Btw, a quick prototype for a managed `JIT_GetRuntimeType`: https://github.com/EgorBo/runtime-1/commit/ef7a46583145b9917865dc1eb1cc85ccc041588e we can enable inlining for CastHelpers and get faster `typeof` for generics (if it shows any benefit)

@jakobbotsch does the jit part look good in this PR?

EgorBo commented:
/azp run runtime-coreclr outerloop, gcstress0x3-gcstress0xc

jkotas commented:
> Btw, a quick prototype for a managed JIT_GetRuntimeType: https://github.com/EgorBo/runtime-1/commit/ef7a46583145b9917865dc1eb1cc85ccc041588e

We will get to this eventually. I think we need to do more cleanup of the runtime structures to make this sort of optimizations interesting and easy to maintain. For example, the writeable data indirection is useless. It should be inlined into Methodtable.

jakobbotsch commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1615,45 +1615,19 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
-    // make this bulletproof, so it works even if we are wrong.
-    if (ReadProcessMemory(GetCurrentProcess(), (void*)strHandle, buff, 4, nullptr) == 0)
+    const int maxStrSize      = 128;
+    char16_t  str[maxStrSize] = {0};
```
```suggestion
    char16_t  str[maxStrSize];
```
Nit, seems unnecessary. Feel free to ignore.

jakobbotsch commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -4874,6 +4897,67 @@ int MethodContext::repGetStringLiteral(CORINFO_MODULE_HANDLE module, unsigned me
+        {
+            char16_t* srcBuffer = (char16_t*)ObjectToString->GetBuffer(value.B);
+            Assert(srcBuffer != nullptr);
+            memcpy(buffer, srcBuffer, min(srcBufferLength, bufferSize) * sizeof(char16_t));
```
Looks like this is missing null termination too

EgorBo replied:
doesn't it copy already zero terminated array so it's not needed?

jakobbotsch replied:
`srcBufferLength` is the recorded count of characters IIUC. And I guess it does not handle the empty string case

EgorBo replied:
and so is `srcBuffer` also recorded. So since empty string is handled in VMs is this still needed?

EgorBo replied:
Removed null-termination from API to match existing behavior of getStringLiteral

jakobbotsch commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -715,6 +715,52 @@ int CEEInfo::getStringLiteral (
+        ((ReflectClassBaseObject*)obj)->GetType().GetName(stackStr);
+    }
+
+    if (stackStr.GetCount() > 0)
+    {
+        result = min(bufferSize, (int)stackStr.GetCount());
+        memcpy((BYTE*)buffer, stackStr.GetUnicode(), result * 2);
+        // Null-terminate it (trim if needed)
+        buffer[result == bufferSize ? result - 1 : result] = 0;
+    }
```
This is still missing null termination for the empty string case.

EgorBo replied:
@jakobbotsch the buffer is zeroed on jit side so technically it's not but I assume it's better to handle it here

jakobbotsch replied:
I think so too. Otherwise it makes the contract pretty strange/inefficient.

EgorBo replied:
Fixed, thanks!

EgorBo replied:
checked that it works:
```asm
       48B80800AD1144020000 mov      rax, 0x24411AD0008      ; ''
```

EgorBo committed:
Update src/coreclr/jit/ee_il_dll.cpp

Co-authored-by: Jakob Botsch Nielsen <Jakob.botsch.nielsen@gmail.com>

jakobbotsch commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1615,45 +1615,40 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
     }
 
-    if (asString->stringLen >= 255 || asString->chars[asString->stringLen] != 0)
+    if (realLength >= 0)
```
Looks like this case needs to be moved up now

jakobbotsch commented on src/coreclr/inc/corinfo.h:
```diff
@@ -2265,6 +2265,13 @@ class ICorStaticInfo
             int                         bufferSize  /* IN  */
             ) = 0;
 
+    // Calls ToString() for given pinned/frozen object handle
```
May be worth to document that the buffer is not null terminated.

EgorBo replied:
Done

EgorBo committed:
remove null-termination from api

jkotas commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1615,45 +1615,40 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
+    if (realLength >= 0)
     {
-        return nullptr;
+        printf("%s '%S'\n", prefix, str);
```
It would be better to return the string from the JIT/EE interface API as utf8, similar to other 

The problem with depending on `%S` to format `char16_t` strings is that depends on the PAL emulating Windows CRT behavior on Unix. We have been working towards getting rid of the Windows PAL emulator, and adding more uses of %S goes against that.

EgorBo replied:
@jkotas I was thinking of leaving it as is just for future JIT optimizations e.g. enable unrolling of Equals/StartsWith for static readonly strings, folding some `.ToString()` operations etc and for that I'll need raw content

jkotas replied:
If you want to do that, it should be a new API that just returns the raw string content.

I do not think it is a good idea to be overloading APIs to do both pretty printing and real functionality.

EgorBo replied:
Replaced with UTF8 `char*`
The only problem that it now prints garbage for non latin1 string literals for JitDisasm

jkotas replied:
The same issue exists in number of other places (e.g. for method names). I do not think that it is a problem. Do you agree?

I think we should standardize on utf8 output, and tell everybody who is not happy about the garbage characters from jit tracing to switch their console encoding to utf8. I have my console encoding set to utf8 on all my Windows systems for a while, so I almost forgot that this problem exists.

jkotas commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1615,45 +1615,40 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
+    }
+    else
+    {
+        printf("%s 'frozen object handle'\n", prefix, str);
```
```suggestion
        printf("%s 'frozen object handle'\n", prefix);
```
The formatting string is only using a single arg.

EgorBo committed:
address feedback

EgorBo commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1617,9 +1617,30 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
+
+    for (size_t i = 0; i < min(maxStrSize, realLength); i++)
+    {
+        // Escape \n and \r symbols
```
Will change the comment, it's needed because `\n` is propagated into JitDisasm and ruins output

jakobbotsch commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1615,45 +1615,35 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
-    if (impGetStringClass() == *((CORINFO_CLASS_HANDLE*)strHandle))
+    const int maxStrSize = 64;
+    char16_t  str[maxStrSize];
+    int       realLength = this->info.compCompHnd->objectToString((void*)handle, str, maxStrSize);
```
I think we still need to handle -1 return here since that can come from SPMI on missing data.

EgorBo replied:
Fixed, thanks!

EgorBo committed:
Address feedback

EgorBo committed:
Pass utf8 string in objectToString

EgorBo committed:
remove char16_t left over

EgorBo committed:
fix compilation warning

jakobbotsch commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -6911,9 +6911,9 @@ GenTree* Compiler::gtNewIndOfIconHandleNode(var_types indType, size_t addr, GenT
+        if ((iconFlags == GTF_ICON_STR_HDL) || (iconFlags == GTF_ICON_OBJ_HDL))
         {
-            // String literals are never null
+            // String literals and frozen type objects are never null
```
Seems like this is dead code? This would give method table. I don't think we would go through this path for frozen objects even when getting method table.

EgorBo replied:
Good point, will remove

jakobbotsch commented on src/coreclr/jit/morph.cpp:
```diff
@@ -11161,7 +11181,7 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
             if (!tree->AsIndir()->IsVolatile())
             {
-                if (op1->IsIconHandle(GTF_ICON_STR_HDL))
+                if (op1->IsIconHandle(GTF_ICON_OBJ_HDL))
```
I guess we see this for cases like `"" is string`?

EgorBo replied:
I bet this can be removed since we now add these flags in morph anyway in `case GT_IND:` and the `is volatile` for string literal load seems to be useless - I'll check it in a follow up together with https://github.com/dotnet/runtime/pull/75573/files#r977367516

jakobbotsch commented on src/coreclr/jit/optcse.cpp:
```diff
@@ -812,7 +812,7 @@ bool Compiler::optValnumCSE_Locate()
                         // Unconditionally allow these constant handles to be CSE'd
                         !tree->IsIconHandle(GTF_ICON_STATIC_HDL) && !tree->IsIconHandle(GTF_ICON_CLASS_HDL) &&
-                        !tree->IsIconHandle(GTF_ICON_STR_HDL))
+                        !tree->IsIconHandle(GTF_ICON_STR_HDL) && !tree->IsIconHandle(GTF_ICON_OBJ_HDL))
```
Doesn't `GTF_ICON_STR_HDL` come with an invariant indirection on top? Do you know why we would CSE the constant but not the indir (now and also before)?

EgorBo replied:
@jakobbotsch normally it's expected that indirs will be CSE'd, unlike raw constants they have higher weight and CSE should handle them on x64, I don't remember why exactly I put `GTF_ICON_STR_HDL` here in that other PR but AFAIR it affected diffs so I left it (initially I only wanted to handle `GTF_ICON_CLASS_HDL`) - I'll check diffs after this PR what will happen if I remove `GTF_ICON_STR_HDL` from here

EgorBo replied:
Ah, good point, will check

jakobbotsch reviewed (approved):
JiT changes LGTM, a few suggestions and questions.

EgorBo committed:
Update gentree.cpp

EgorBo commented:
Looks like this can be merged, two Pri0 failures are from https://github.com/dotnet/runtime/pull/75952, SPMI are due to JIT-EE guid change (we still don't know how to avoid triggering those when both JIT and JIT-EE guid are changed in YML)

I plan to look into the profiler/diag issues raised in the other PR for FOH, and test this and the string literals on OSS (and, hopefully, some 1st parties) projects to get a better understanding how many objects make it to FOH.

Then we might look into making `static` objects (at least those without GC fields inside) also FOH-based

jkotas commented:
> Then we might look into making static objects (at least those without GC fields inside) also FOH-based

How do you plan to do that? I do not think it is feasible without AOT step (or without doing expensive work when JITing static constructors - I do not expect we would want to do that).

jakobbotsch commented:
Can we do it for static field boxes (the ones created for non-primitives)? Today, JIT needs to create
```scala
ADD byref
  IND  ref
    CNS <object slot addr>
  CNS 8
```
for all such accesses, I believe. We could potentially create direct pointer to the boxed contents.

jkotas commented:
Ah ok, yes - these boxes can be allocated on the frozen heap.

EgorBo commented:
Good point about the static boxes, regarding other types of objects I am not sure either

Perhaps, for e.g. 
```csharp
static readonly string str = ...;
```
we could insert `call object CopyToFoh(object o)` call before `stsfld` in `.cctor` as people, in theory, should not rely on references for string objects but it probably sounds terrible

EgorBo commented:
Improvements https://github.com/dotnet/perf-autofiling-issues/issues/8700
